### DISCUSSION TITLE
Histogram - reduction using Expr instead of Var

### DISCUSSION THREAD

**Discussion by timeinpixels** [halide/Halide#6332] [Q&A] [answered]
I was playing a bit with the histogram implementation in Halide and found this problem... 
It is actually an RGB waveform rather than a histogram but took the initial idea from the histogram example.
Is there a way around it?

Simplified code snippet:

```
    Halide::Func waveform, result;

    Halide::Var x, y, c;
    Halide::RDom r(0, input.width(), 0, input.height());

    waveform(x, y, c) = 0;
    waveform(x, input(r.x, r.y, 0), c) = Halide::cast<int>(input(x, y, c));

    result(x, y, c) = Halide::cast<uint8_t>(waveform(x, y, c));

    Halide::Buffer<uint8_t> output =
        result.realize({input.width(), input.height(), input.channels()});

```

I am getting:
`libc++abi.dylib: terminating with uncaught exception of type Halide::CompileError: Error: Undefined variable "v1" in definition of Func "f0"`

I tried allocating an intermediate buffer, convert the input image to the luminosity (1 channel) buffer, and pass it as 'y' coord to histogram, but that threw an exception as well. 
What should be the right way to do it?

**Comment by abadams** [comment, marked as answer]
y doesn't exist on the left-hand-side of your update definition for waveform, so I'm not sure which pixel you're trying to load. Perhaps you mean the following: ?

```
    Halide::Func waveform, result;

    Halide::Var x, y, c;
    Halide::RDom r(0, input.width(), 0, input.height());

    waveform(x, y, c) = 0;
    waveform(r.x, input(r.x, r.y, 0), c) = Halide::cast<int>(input(r.x, r.y, c));

    result(x, y, c) = Halide::cast<uint8_t>(waveform(x, y, c));

    Halide::Buffer<uint8_t> output =
        result.realize({input.width(), input.height(), input.channels()});
```

**Comment by timeinpixels** [comment]
Oh, makes sense! Thanks!
